// unity-api/dist/shortcuts.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

var createShortcut = function createShortcut(method) {
    return function () {
        var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        var _ref$options = _ref.options,
            options = _ref$options === undefined ? {} : _ref$options,
            rest = _objectWithoutProperties(_ref, ['options']);

        return _extends({}, rest, {
            options: _extends({
                method: method
            }, options)
        });
    };
};

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods
var GET = exports.GET = createShortcut('GET');
var HEAD = exports.HEAD = createShortcut('HEAD');
var POST = exports.POST = createShortcut('POST');
var PUT = exports.PUT = createShortcut('PUT');
var DELETE = exports.DELETE = createShortcut('DELETE');
var CONNECT = exports.CONNECT = createShortcut('CONNECT');
var OPTIONS = exports.OPTIONS = createShortcut('OPTIONS');
var TRACE = exports.TRACE = createShortcut('TRACE');
var PATCH = exports.PATCH = createShortcut('PATCH');
