// rmc-picker/es/NativePicker.android.js
import React from 'react';
import { ScrollView, View, StyleSheet, PixelRatio, Text } from 'react-native';
import createClass from 'create-react-class';
import PickerMixin from './PickerMixin';
var ratio = PixelRatio.get();
var styles = StyleSheet.create({
    indicator: {
        position: 'absolute',
        left: 0,
        top: -99,
        borderColor: '#aaa',
        borderTopWidth: 1 / ratio,
        borderBottomWidth: 1 / ratio
    },
    scrollView: {
        height: 0
    },
    selectedItemText: {
        fontSize: 20,
        fontWeight: 'bold',
        color: '#000'
    },
    itemText: {
        fontSize: 20,
        color: '#aaa',
        textAlign: 'center'
    }
});
var Picker = createClass({
    mixins: [PickerMixin],
    statics: {
        Item: function Item() {}
    },
    onItemLayout: function onItemLayout(e) {
        var _this = this;

        var _e$nativeEvent$layout = e.nativeEvent.layout,
            height = _e$nativeEvent$layout.height,
            width = _e$nativeEvent$layout.width;
        // console.log('onItemLayout', height);

        if (this.itemHeight !== height || this.itemWidth !== width) {
            this.itemWidth = width;
            this.refs.indicator.setNativeProps({
                style: [styles.indicator, {
                    top: height * 3,
                    height: height,
                    width: width
                }]
            });
        }
        if (this.itemHeight !== height) {
            this.itemHeight = height;
            this.refs.scroller.setNativeProps({
                style: {
                    height: height * 7
                }
            });
            this.refs.content.setNativeProps({
                style: {
                    paddingTop: height * 3,
                    paddingBottom: height * 3
                }
            });
            // i do no know why!...
            setTimeout(function () {
                _this.select(_this.props.selectedValue);
            }, 0);
        }
    },
    componentDidUpdate: function componentDidUpdate() {
        this.select(this.props.selectedValue);
    },
    componentWillUnMount: function componentWillUnMount() {
        this.clearScrollBuffer();
    },
    clearScrollBuffer: function clearScrollBuffer() {
        if (this.scrollBuffer) {
            clearTimeout(this.scrollBuffer);
        }
    },
    scrollTo: function scrollTo(y) {
        this.refs.scroller.scrollTo({
            y: y,
            animated: false
        });
    },
    fireValueChange: function fireValueChange(selectedValue) {
        if (this.props.selectedValue !== selectedValue && this.props.onValueChange) {
            this.props.onValueChange(selectedValue);
        }
    },
    onScroll: function onScroll(e) {
        var _this2 = this;

        var y = e.nativeEvent.contentOffset.y;

        this.clearScrollBuffer();
        this.scrollBuffer = setTimeout(function () {
            _this2.clearScrollBuffer();
            _this2.doScrollingComplete(y);
        }, 100);
    },
    render: function render() {
        var _this3 = this;

        var _props = this.props,
            children = _props.children,
            itemStyle = _props.itemStyle,
            selectedValue = _props.selectedValue,
            style = _props.style;

        var items = React.Children.map(children, function (item, index) {
            var totalStyle = [styles.itemText];
            if (selectedValue === item.props.value) {
                totalStyle.push(styles.selectedItemText);
            }
            totalStyle.push(itemStyle);
            return React.createElement(
                View,
                { ref: 'item' + index, onLayout: index === 0 ? _this3.onItemLayout : undefined, key: item.key },
                React.createElement(
                    Text,
                    { style: totalStyle, numberOfLines: 1 },
                    item.props.label
                )
            );
        });
        return React.createElement(
            View,
            { style: style },
            React.createElement(
                ScrollView,
                { style: styles.scrollView, ref: 'scroller', onScroll: this.onScroll, scrollEventThrottle: 16, showsVerticalScrollIndicator: false, overScrollMode: 'never' },
                React.createElement(
                    View,
                    { ref: 'content' },
                    items
                )
            ),
            React.createElement(View, { ref: 'indicator', style: styles.indicator })
        );
    }
});
export default Picker;
